### PR TITLE
Homework1

### PR DESCRIPTION
**Pull request by violonistahiles**
First try to solve homework1.

### CODE REVIEW

File: homework1/sample_project/calculator/calc.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+def check_power_of_2(a: int) -> bool:
+
+    if a == 0:
+        return False
+
+    result = not (bool(a & (a - 1)))
```
Comment (icanhardcode) [inline homework1/sample_project/calculator/calc.py]: u can return result without creating additional variable

File: homework1/task02.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+            if k == data_len:
+                return True
+            else:
+                k += 1
+        else:
+            if k > 0:
+                return False
+
+        fib_sequence.append(sum(fib_sequence))
+        fib_sequence = fib_sequence[1:]
```
Comment (icanhardcode) [inline homework1/task02.py]: you can do this process by another way.. try to think

File: homework1/task02.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    and returns if the given sequence is a Fibonacci sequence
+
+We guarantee, that the given sequence contain >= 0 integers inside.
+
+"""
+from typing import Sequence
+
+
+def check_fibonacci(data: Sequence[int]) -> bool:
+
```
Comment (nikitosboroda) [inline homework1/task02.py:13]: your squence should be length of 3 to be the fibb

File: homework1/task04.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+                value = (first if first[0] <= second[0] else second).pop(0)
+                if len(first) > 0 and value == first[0]:
+                    first.pop(0)
+                    yield value
+
+                if len(second) > 0 and value == second[0]:
+                    second.pop(0)
+                    yield value
+
+        return list(_take_equal())
```
Comment (icanhardcode) [inline homework1/task04.py]: u did extra work.. + pop(0) - O(N) complexity

File: homework1/task04.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+    # Divide sum of four elements to two sums of two elements
+    # Get sums of the first two lists
+    subset_ab = defaultdict(int)
+    for i in range(len(a)):
+        for j in range(len(b)):
+            subset_ab[a[i] + b[j]] += 1
+    ab_values = sorted(list(subset_ab.keys()))
```
Comment (icanhardcode) [inline homework1/task04.py]: u dont need this

File: homework1/task04.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+    # Get sums of the second two lists with minus
+    subset_cd = defaultdict(int)
+    for i in range(len(c)):
+        for j in range(len(d)):
+            subset_cd[-(c[i] + d[j])] += 1
+    cd_values = sorted(list(subset_cd.keys()))
```
Comment (icanhardcode) [inline homework1/task04.py]: and this

File: homework1/task04.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+    # Select only equal elements from two lists
+    valid_values = take_equal(ab_values, cd_values)
+    result = 0
+    # According to the number of sum entries get final tuples number
+    for key in valid_values:
+        result += subset_cd[key] * subset_ab[key]
```
Comment (icanhardcode) [inline homework1/task04.py]: extra work.. find out another approach

File: homework1/task05.py [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    result = 16
+"""
+from typing import List
+
+
+def find_maximal_subarray_sum(nums: List[int], k: int) -> int:
+
+    nums_len = len(nums)
+    if k > nums_len:
+        k = nums_len
```
Comment (icanhardcode) [inline homework1/task05.py]: should return None or raise exception

File: tests/homework1/test_task05.py [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+
+def test_nums_contain_zero_elements_case():
+    """Testing program with no elements in nums list"""
+    nums_test = []
+    k = 3
+
+    result = find_maximal_subarray_sum(nums_test, k)
+
+    assert result is None
```
Comment (icanhardcode) [inline tests/homework1/test_task05.py]: u dont need ```is None```.. there is another way

File: tests/homework1/test_task02.py [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+def test_not_a_fib_sequence_case():
+    """Testing that not a Fibonacci sequence give False"""
+    data_to_process = [0, 1, 1, 7]
+    assert not check_fibonacci(data_to_process)
+
+
+def test_single_value_from_fib_sequence_case():
+    """Testing that single number from actual Fibonacci sequence give True"""
+    data_to_process = [2584]
+    assert check_fibonacci(data_to_process)
```
Comment (icanhardcode) [inline tests/homework1/test_task02.py]: regarding comment in task file.. that incorrect sequence

File: homework1/task03.py [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+        # Process other lines and compare values
+        for line in file_example:
+            number = int(line.strip())
+
+            if number < min_value:
+                min_value = number
+            if number > max_value:
+                max_value = number
+
+    return min_value, max_value
```
Comment (icanhardcode) [inline homework1/task03.py]: that not enough clear logic... take a look at functions name and result which it returns

**Review by icanhardcode — CHANGES_REQUESTED** [review]

File: homework1/sample_project/calculator/calc.py (hunk 1 above)
Comment (violonistahiles) [reply homework1/sample_project/calculator/calc.py]: Fixed

File: homework1/task02.py (hunk 2 above)
Comment (violonistahiles) [reply homework1/task02.py]: Changed solution little bit

File: homework1/task02.py (hunk 3 above)
Comment (violonistahiles) [reply homework1/task02.py:13]: Fixed

File: homework1/task04.py (hunk 7 above)
Comment (violonistahiles) [reply homework1/task04.py]: Did it in different way

File: homework1/task05.py (hunk 8 above)
Comment (violonistahiles) [reply homework1/task05.py]: Fixed

File: tests/homework1/test_task05.py (hunk 9 above)
Comment (violonistahiles) [reply tests/homework1/test_task05.py]: Fixed

File: tests/homework1/test_task02.py (hunk 10 above)
Comment (violonistahiles) [reply tests/homework1/test_task02.py]: Fixed

File: homework1/task03.py (hunk 11 above)
Comment (violonistahiles) [reply homework1/task03.py]: Fixed

File: homework1/task02.py [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+    fib_sequence_limit = 3
```
Comment (icanhardcode) [inline homework1/task02.py]: dont createadditional variables if you can handle without them

File: homework1/task02.py [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+    k = 0
+    while data[k] >= fib_1:
+        is_equal = (data[k] == fib_1)
+
+        if is_equal and k == data_len:
+            return True
+        elif is_equal and k < data_len:
+            k += 1
+        elif k > 0:
+            return False
```
Comment (icanhardcode) [inline homework1/task02.py]: ```if ... else``` cause u do return in previous if

**Review by icanhardcode — CHANGES_REQUESTED** [review]

**Review by icanhardcode — APPROVED** [review]

**Pull request merged**